sharding/indexHashedNodesCoordinatorWithRater.go

```diff
@@ -76,76 +73,28 @@ func (ihgs *indexHashedNodesCoordinatorWithRater) IsInterfaceNil() bool {
 	return ihgs == nil
 }
 
-func (ihgs *indexHashedNodesCoordinatorWithRater) expandAllLists(epoch uint32) error {
-	ihgs.mutNodesConfig.Lock()
-	defer ihgs.mutNodesConfig.Unlock()
-
-	nodesConfig, ok := ihgs.nodesConfig[epoch]
-	if !ok {
-		return ErrEpochNodesConfigDesNotExist
-	}
-
-	shardsExpanded := make(map[uint32][]Validator)
-
-	err := ihgs.expandListsForEpochConfig(nodesConfig, shardsExpanded)
-	if err != nil {
-		return err
-	}
-
-	return nil
-}
-
-func (ihgs *indexHashedNodesCoordinatorWithRater) expandListsForEpochConfig(nodesConfig *epochNodesConfig, shardsExpanded map[uint32][]Validator) error {
-	nodesConfig.mutNodesMaps.Lock()
-	defer nodesConfig.mutNodesMaps.Unlock()
-
-	nodesConfig.expandedEligibleMap = make(map[uint32][]Validator)
-
-	nrShards := len(nodesConfig.eligibleMap)
-	metaChainExpanded, err := ihgs.expandList(nodesConfig, core.MetachainShardId)
-	if err != nil {
-		return err
-	}
-
-	for shardId := uint32(0); shardId < uint32(nrShards-1); shardId++ {
-		shardsExpanded[shardId], err = ihgs.expandList(nodesConfig, shardId)
-		if err != nil {
-			return err
-		}
-	}
-
-	nodesConfig.expandedEligibleMap[core.MetachainShardId] = metaChainExpanded
-	for shardId := uint32(0); shardId < uint32(nrShards-1); shardId++ {
-		nodesConfig.expandedEligibleMap[shardId] = shardsExpanded[shardId]
-	}
-	return nil
-}
-
-func (ihgs *indexHashedNodesCoordinatorWithRater) expandList(nodesConfig *epochNodesConfig, shardId uint32) ([]Validator, error) {
-	validators := nodesConfig.eligibleMap[shardId]
-	log.Trace("Expanding eligible list", "shardID", shardId)
-	return ihgs.expandEligibleList(validators)
-}
-
-func (ihgs *indexHashedNodesCoordinatorWithRater) expandEligibleList(validators []Validator) ([]Validator, error) {
+// ValidatorsWeights returns the weights/chances for each given validator
+func (ihgs *indexHashedNodesCoordinatorWithRater) ValidatorsWeights(validators []Validator) ([]uint32, error) {
 	minChance := ihgs.GetChance(0)
-	minSize := len(validators) * int(minChance)
-	validatorList := make([]Validator, 0, minSize)
+	weights := make([]uint32, len(validators))
 
-	for _, validatorInShard := range validators {
+	loadingFromDisk := ihgs.loadingFromDisk.Load().(bool)
+	for i, validatorInShard := range validators {
 		pk := validatorInShard.PubKey()
-		rating := ihgs.GetRating(string(pk))
-		chances := ihgs.GetChance(rating)
-		if chances < minChance {
-			//default chance if all validators need to be selected
-			chances = minChance
-		}
-		log.Trace("Computing chances for validator", "pk", pk, "rating", rating, "chances", chances)
-
-		for i := uint32(0); i < chances; i++ {
-			validatorList = append(validatorList, validatorInShard)
+		if !loadingFromDisk {
```

Comment:
shouldn't we try multiple times until we don't read anymore from disk ? or it is ok to just skip this entry if reading from disk is in progress

Reply:
in case this function is called because we are restoring configuration from disk we already have the weights calculated, so we do not need to get them from trie and we just load them with the other parameters, otherwise we read the rating from trie and compute the associated chances/weights